### PR TITLE
Trie routing

### PR DESCRIPTION
**Pull request by robhobbes**
- Adding some comments about what still needs to be done
- Switching to `SmallIntMap` instead of `HashMap`
- Cleaned things up a bit
  - Renamed `TrieNode` to `TrieRouterNode`
  - Gave `TrieRouter` responsibility for building the trie
  - Made `new` the constructor for `TrieRouterNode` instead of `setup`

### CODE REVIEW

File: src/route/trierouter.rs [hunk 1]
```diff
@@ -124,19 +146,71 @@ impl<'a> TrieRouter<'a> {
...
+            method::Get => {
+                let mut file = File::open(&Path::new("."+path));
+                let mut result = file.read_to_end();
+                match result {
+                    Ok(_) => {
+                        let pieces: ~[&str] = path.rsplitn('.',1).collect();
+                        let extension = pieces[0];
+                        let content_type = mimetype::content_type_from_ext(extension);
+                        response.headers.content_type = Some(content_type);
+                        response.write(result.unwrap());
```
Comment (jroweboy) [inline src/route/trierouter.rs]: instead of `result.unwrap()` you could just change the `Ok(_)` line to say `Ok(res)` since that is unwrapping the contents right there and allowing you to use them in the match block and then you can just `response.write(res)`

File: src/route/trierouter.rs [hunk 2]
```diff
@@ -232,17 +278,64 @@ impl TrieNode {
...
+    }
+}
+
+// HashMap has this great find_or_insert function
+// but SmallIntMap doesn't, so I it
+pub trait FindOrInsert {
+    fn find_or_insert<'a>(&'a mut self, ch: char, node: TrieRouterNode) -> &'a mut TrieRouterNode;
+}
+
+impl FindOrInsert for SmallIntMap<TrieRouterNode> {
```
Comment (jroweboy) [inline src/route/trierouter.rs:336]: We should add this method to the rust source code since it seems kinda silly that it already isn't there

File: src/mimetype.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+pub fn content_type_from_ext(extension: &str) -> MediaType {
+    let pieces: ~[&str] = mimetype_from_ext(extension).split('/').collect();
+    MediaType {
+        type_: pieces[0].to_owned(),
+        subtype: pieces[1].to_owned(),
+        parameters: Vec::new(),
+    }
+}
+
+fn mimetype_from_ext<'a>(extension: &str) -> ~str {
```
Comment (jroweboy) [inline src/mimetype.rs]: This function probably (I'm guessing) doesn't need the lifetime specifier since you aren't using it anywhere

File: src/route/trierouter.rs [hunk 4]
```diff
@@ -92,8 +110,8 @@ impl<'a> TrieRouter<'a> {
             if building_var && ch == '/' {
                 vars.push((current_key.clone(),current_var.clone()));
                 building_var = false;
-                if current_node.children.contains_key(&ch) {
-                    current_node = current_node.children.get(&ch);
+                if current_node.children.contains_key(&(ch as uint)) {
```
Comment (jroweboy) [inline src/route/trierouter.rs:118]: Just a personal opinion, but I think `&ch as &uint` looks a lot better than `&(ch as uint)` since this first makes me think that you are taking a reference to a `char` and casting the reference as a `&uint` instead of the way you have it written where it looks like you are casting the `char` to a `uint` and then taking a reference to that. Not that they are actually different, its just how I parse them. 

In short, I'd love it to say `&ch as &uint` instead :smile:

File: src/route/trierouter.rs (hunk 2 above)
Comment (robhobbes) [inline src/route/trierouter.rs:336]: Agreed... but I'm not gonna do it right now. I don't know how to do it generically.

File: src/route/trierouter.rs (hunk 4 above)
Comment (robhobbes) [inline src/route/trierouter.rs:118]: I don't believe I can do it that way. You'll have to play with it and see if it is possible. When I do `&ch as &uint` it says something about "non-scalar cast"

**Comment by jroweboy** [comment]
Changed to using Ogeon's router instead. One day after doing some benchmarking though I'll see which is faster. I decided to make the router code part of the framework since that is one of the essential parts to any website so I removed the Router Trait but still left it modular enough to be swapped out if I find a better faster router.
